File: app/src/main/java/com/limelight/ligase/stream/StreamSessionExitCoordinator.java
package com.limelight.ligase.stream;

import java.util.concurrent.atomic.AtomicBoolean;

public final class StreamSessionExitCoordinator {
    public interface Callbacks {
        void disconnect();
        void endSession();
    }

    private final AtomicBoolean exitRequested = new AtomicBoolean();
    private final Callbacks callbacks;

    public StreamSessionExitCoordinator(Callbacks callbacks) {
        this.callbacks = callbacks;
    }

    public boolean disconnect() {
        if (!exitRequested.compareAndSet(false, true)) {
            return false;
        }
        callbacks.disconnect();
        return true;
    }

    public boolean endSession() {
        if (!exitRequested.compareAndSet(false, true)) {
            return false;
        }
        callbacks.endSession();
        return true;
    }

    public boolean isExitRequested() {
        return exitRequested.get();
    }
}

File: app/src/main/java/com/limelight/ligase/stream/StreamSessionExitCoordinator.kt
package com.limelight.ligase.stream

import java.util.concurrent.atomic.AtomicBoolean

class StreamSessionExitCoordinator(
    private val callbacks: Callbacks,
) {
    interface Callbacks {
        fun disconnect()
        fun endSession()
    }

    private val exitRequested = AtomicBoolean()

    fun disconnect(): Boolean {
        if (!exitRequested.compareAndSet(false, true)) {
            return false
        }
        callbacks.disconnect()
        return true
    }

    fun endSession(): Boolean {
        if (!exitRequested.compareAndSet(false, true)) {
            return false
        }
        callbacks.endSession()
        return true
    }

    fun isExitRequested(): Boolean = exitRequested.get()
}
